server/src/test/java/io/crate/planner/optimizer/iterative/IterativeOptimizerTest.java

```diff
@@ -121,5 +125,25 @@ public void test_reapply_rules_on_changing_nodes() {
         );
     }
 
+    @Test
+    public void test_optimizer_is_interrupted_when_exceeds_timeout() {
```

Comment:
The alternative is to add a rule which a wait which exceeds the timeout token.

Reply:
Oh i think this would not exceed the timeout on the optimizer since it checked by iterations.

Reply:
Is it okay to keen nested order then? Seems like less effort (and tests actual production rules)

Reply:
I would match on a single Filter many times instead of adding many plans, see the updated rule example.

Reply:
Token won't be checked without hitting 100 transformations.  

It will timeout with the slow rule but because we check it in the Session after `planner.plan `- so it would be somewhat overlapping with the existing tests

Reply:
```
    @Test
    public void test_optimizer_is_interrupted_when_exceeds_timeout() {
        var plan = new MemoTest.TestPlan(1, List.of());
        IterativeOptimizer optimizer = new IterativeOptimizer(nodeCtx,
            () -> Version.CURRENT,
            List.of(new SlowRule()));

        Session.TimeoutToken token = new Session.TimeoutToken(TimeValue.timeValueMillis(1), System.nanoTime());
        assertThatThrownBy(() -> optimizer.optimize(plan, planStats, ctx, OptimizerTracer.NOOP, token)
        )
            .isExactlyInstanceOf(JobKilledException.class)
            .hasMessage("Job killed. statement_timeout (1)");
    }

    static class SlowRule implements Rule<MemoTest.TestPlan> {

        @Override
        public Pattern<MemoTest.TestPlan> pattern() {
            return Pattern.typeOf(MemoTest.TestPlan.class);
        }

        @Override
        public LogicalPlan apply(MemoTest.TestPlan plan, Captures captures, Context context) {
            try {
                Thread.sleep(Duration.ofMillis(10));
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            return plan;
        }
    }
```

Reply:
This works.

Reply:
Oh right, we return non null in apply so we will transform. Applied the suggestion, thanks!